Introduces Prefix enum

brigand opened the pull request:
Hey, this is my first rust PR.  Very open to making any changes.

Ref [#143 Parsing message prefixes (sources) into an enumeration](https://github.com/aatxe/irc/issues/143)

---

There's an issue in that both of the following are allowed by the types, but there's only one string representation for both of them. The parser never produces this (gives the second), so it'd just be manual enum construction.

```rust
Nickname("nick".into(), None, Some("host".into()))
Nickname("nick".into(), Some(String::new()), Some("host".into()))
```

This is harmless until you try to do...

```
let prefix = Prefix::Nickname("nick".into(), None, Some("host".into()));
let s = format!("{}", prefix);
assert_eq(prefix, s.parse.unwrap()); // panics
```

---

I'm not sure what this function should be named, but it seems weird doing `string.parse().unwrap()` everywhere, since it's not apparent at the call site that it never fails.

```rust
impl Prefix {
   pub fn new_from_str(s: &str) -> Prefix {
```

Edit: not sure which branch this should be based on.

brigand committed:
Prefix with FromStr and Display

brigand committed:
proto::message uses Prefix

brigand committed:
prefix cleanup and docs

aatxe commented:
Thanks for the pull request! :beers:

Before I do a more in-depth review (mostly because I don't have a ton of time at this moment), let's knock out some higher-level things. #148 was merged already, so you should be able to rebase (and maybe have to do some git fiddling?) to get rid of 79e00bf. That being said, since this is a breaking change, it would be preferable to have it based on `0.14` instead of `develop`. So, to that end, I've updated `0.14` to bring it inline with all the changes made on develop lately (since I haven't been super actively working on `0.14`). You should be able to base it on `0.14`, but please let me know if you have any problems!

Afterward, I'll review the code in more depth. 😄

8573 commented:
> There's an issue in that both of the following are allowed by the types, but there's only one string representation for both of them. The parser never produces this (gives the second), [...]

Per the IETF RFC 2812 grammar that #143 cites, the `user` and `host` components of `prefix`es, while optional, may not be the empty string. I would suggest not having multiple ways to indicate that such a component is absent, but rather either
1. representing the `user` and `host` components as `String` rather than `Option<String>`, and using the empty string as the sole means of indicating that a component was absent; or
2. representing these components as `Option<S>` where `S` is some string type that can't be the empty string.

(I suppose this is more a comment on the design in #143 than its implementation in these patches, but this ticket is where the action is.)

A bit of searching on Crates.io doesn't show me any implementations of a non-empty string type, though, so I'd favor (1).

brigand committed:
rebased

brigand committed:
prefix uses Nickname(String, String, String)

brigand committed:
prefix docs

brigand commented:
Thanks for the feedback!

@8573 I agree, None and the empty string are semantically the same here. Changed over to that, which also simplified the parser.

@aatxe rebased on 0.14, and it's ready for review.

8573 commented on irc-proto/src/prefix.rs:
```diff
@@ -0,0 +1,204 @@
+        enum Active {
+            Name = 0,
+            User = 1,
+            Host = 2,
```
In Rust, we (usually) only specify numeric values for enum variants if we need to refer to them not by name but by number, such as from other languages such as C.

brigand replied:
Oops, was left over from indexing into a `[String; 3]`. Fixed.

aatxe commented:
@8573 hehe, I'll cover my tracks by saying that I said "probably" 😉, but I agree that you've got the right idea. So, I'm glad @brigand followed.

@brigand Thanks for the rebase!

It looks like some tests are failing on Travis:
```rust
test prefix::test::parse_danger_cases ...FAILED
test message::test::from_and_to_string ...ok
test message::test::new ...ok
test message::test::source_nickname ...FAILED
test message::test::to_message_invalid_format ...ok
failures:
---- prefix::test::parse_danger_cases stdout ----
thread 'prefix::test::parse_danger_cases' panicked at 'assertion failed: `(left == right)`
  left: `"name!@"`
 right: `"name"`', irc-proto/src/prefix.rs:115:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
---- message::test::source_nickname stdout ----
thread 'message::test::source_nickname' panicked at 'assertion failed: `(left == right)`
  left: `Some("test@test")`
 right: `Some("test")`', irc-proto/src/message.rs:313:9
failures:
    message::test::source_nickname
    prefix::test::parse_danger_cases
```

Do these tests work on your machine? If not, can you fix them?
You may have missed them because they're in `irc-proto`, rather than `irc`.
This distinction only exists on 0.14 right now.

8573 commented on irc-proto/src/prefix.rs:
```diff
@@ -0,0 +1,204 @@
+                (name, user, "") => write!(f, "{}!{}", name, user),
+                // This case shouldn't happen normally, but user!@host is invalid, so we
+                // format it without the host
+                (name, "", _host) => write!(f, "{}", name),
```
`nickname!@hostname` is invalid, but `nickname@hostname` is valid, and wouldn't the latter also result in this match arm being used?

8573 commented on irc-proto/src/prefix.rs:
```diff
@@ -0,0 +1,204 @@
+
+/// This implementation never returns an error and is isomorphic with `Display`.
+impl FromStr for Prefix {
+    type Err = string::ParseError;
```
I think the unit type `()` usually is used when this can't return an error. (Eventually it instead will be the never type `!`.)

brigand replied:
This was recommended to me as it's basically `!`, and `let s: Prefix = s.parse().unwrap()` will continue to be safe unless the `type Err` is changed.

I'd like to keep the current code.

aatxe replied:
`string::ParseError` is definitely not `!` (which should be read as "diverges"/"never returns"/etc.). As it stands, the type says that it might return a `ParseError`, even if you never return it. Ideally, it should be a `!`, but since that is not currently allowed in Rust, the current convention for non-erroring things is to use `()`.

brigand replied:
`string::ParseError` is an empty enum, so there's no way to a value of type `ParseError` in safe rust, which means the function can only possibly return `Ok(s)`.

Changed to `type Err = ()`.

aatxe replied:
Ah, sorry, I missed that part (which is itself an argument to just write `()`, to be honest).

8573 replied:
Now that I've read the docs for `std::string::ParseError` — I didn't realize we had uninhabited error types in `std` — I think it does sound more appropriate than `()`.

8573 commented on irc-proto/src/prefix.rs:
```diff
@@ -0,0 +1,204 @@
+                },
+
+                // The '@' is not special until we've started the username
+                // portion
```
According to the grammar, the hostname component can be present even if the username component is absent, though not vice versa, and `@` would indicate the start of a hostname regardless of the presence or absence of a username. (Usernames can't contain `@`.)

8573 commented on irc-proto/src/prefix.rs:
```diff
@@ -0,0 +1,204 @@
+            match c {
+                // We consider the '.' to be a ServerName except if a ! has already
+                // been encountered.
+                '.' if active == Active::Name => return Prefix::ServerName(s.to_owned()),
```
This would result in `a.b.c!foo@bar` being considered a server name, right? According to the grammar, server names can't include `!` or `@`.

(According to the grammar, nicknames can't include `.`, so `a.b.c!foo@bar` wouldn't be a valid prefix at all, but IRC daemons tend to be much more lenient about nicknames than the grammar is, so I wouldn't be surprised if there are nicknames out there with `.` in them.)

8573 commented:
(My apologies for taking issue with many things in your first Rust PR; don't be discouraged!)

brigand commented:
Oh, hmm, I'm not sure my  tests are running for with `cargo test` locally.

Edit: had to `cd` into `irc-proto`

Have to make some more fixes

brigand committed:
WIP prefix fixes

brigand committed:
fixes issues from review

brigand committed:
ParseError -> ()

brigand commented:
Should have mentioned, this is ready for review.

aatxe commented:
Hi @brigand,

Sorry for the delay. I've looked through everything, and I'm happy with it. This also adds another place where property testing would be useful (#135), but that can come later. Namely, the invariant you describe in the tests (round-tripping a string through Prefix should be lossless) could be tested with random testing.

The code looks good to me! Cheers! 🍻
